src/operator/contrib/bounding_box-inl.cuh

```diff
@@ -57,6 +60,226 @@ int FilterScores(mshadow::Tensor<gpu, 1, DType> out_scores,
+__global__ void nms_apply_kernel_rest(const int topk, int32_t *index,
+                                 const int32_t *batch_start,
+                                 const DType *input,
+                                 const DType *areas,
+                                 const int num, const int stride,
+                                 const int offset_box, const int offset_id,
+                                 const float thresh, const bool force,
+                                 const int encode, const int start_offset,
+                                 const int blocks_per_batch) {
+  constexpr int block_size = 512;
+  const int batch = blockIdx.x / blocks_per_batch;
+  const int start_ref = static_cast<int>(batch_start[batch]) + start_offset;
+  const int block_offset = (N + blockIdx.x % blocks_per_batch) * block_size;
+  const int start = start_ref + block_offset;
+
+  const int size_of_batch = static_cast<int>(batch_start[batch + 1]) - start;
+  const int end = min(size_of_batch, topk - start_offset - block_offset);
+  const int my_pos = start + threadIdx.x;
+  if (threadIdx.x < end && index[my_pos] >= 0) {
+    const int pos_area_offset = static_cast<int>(index[my_pos]);
+    const int pos_offset = pos_area_offset * stride + offset_box;
+    DType my_box[4];
+#pragma unroll
+    for (int i = 0; i < 4; ++i) {
+      my_box[i] = input[pos_offset + i];
+    }
+    const DType my_area = areas[pos_area_offset];
+    int pos_id = 0;
+    if (check_class) {
+      pos_id = static_cast<int>(input[pos_offset - offset_box + offset_id]);
+    }
+
+    for (int ref = start_ref; ref < start_ref + N * block_size; ++ref) {
+      const int ref_area_offset = static_cast<int>(index[ref]);
+      if (ref_area_offset >= 0) {
+        const int ref_offset = ref_area_offset * stride + offset_box;
+        int ref_id = 0;
+        if (check_class) {
+          ref_id = static_cast<int>(input[ref_offset - offset_box + offset_id]);
+          if (ref_id != pos_id) continue;  // different class
+        }
+        DType intersect = Intersect2(input + ref_offset, my_box[0], my_box[2], encode);
+        intersect *= Intersect2(input + ref_offset + 1, my_box[1], my_box[3], encode);
+        const DType iou = intersect /
+          (areas[ref_area_offset] + my_area - intersect);
+        if (iou > thresh) {
+          index[my_pos] = -1;
+          break;
+        }
+      }
+    }
+  }
+}
+
+template<typename DType>
+void NMSApply(mshadow::Stream<gpu> *s,
+              int num_batch, int topk,
+              mshadow::Tensor<gpu, 1, int32_t>* sorted_index,
+              mshadow::Tensor<gpu, 1, int32_t>* batch_start,
+              mshadow::Tensor<gpu, 3, DType>* buffer,
+              mshadow::Tensor<gpu, 1, DType>* areas,
+              int num_elem, int width_elem,
+              int coord_start, int id_index,
+              float threshold, bool force_suppress,
+              int in_format) {
+  using namespace mxnet_op;
+  constexpr int THRESHOLD = 1024;
```

Comment:
one question, why launch bound set to 512 while threshold is 1024?

Reply:
To have bigger register count available and to increase the useful work done by the threads, since I need to keep all the threads around anyway for syncthreads calls. Each thread processes threshold / block_size elements.